### ISSUE TITLE
resulting `ItemsView` isn't narrowed after `isinstance(..., Mapping)` check (and other oddities)

### ISSUE THREAD

**Issue by posita** [astral-sh/ty#3249]
### Summary

[Playground](https://play.ty.dev/181edb3f-355a-465f-b201-cc99380df36a)

```python
def count_the_things(
    *sources: Mapping[T, SupportsInt] | Iterable[tuple[T, SupportsInt]],
    use_keys_instead_of_items_for_maps: bool = False,
) -> Counter[T]:
    c: Counter[T] = Counter()
    for source in sources:
        if isinstance(source, Mapping):
            if use_keys_instead_of_items_for_maps:
                for outcome in source:
                    count = source[outcome]
                    reveal_type(outcome)  # I don't know *where* this is coming from
                    reveal_type(count)  # should be SupportsInt, but is object
                    c[outcome] += int(count)
            else:
                for outcome, count in source.items():
                    reveal_type(outcome)  # should be T, but is object
                    reveal_type(count)  # should be SupportsInt, but is object
                    c[outcome] += int(count)
        else:
            for outcome, count in source:
                reveal_type(outcome)
                reveal_type(count)
                c[outcome] += int(count)
    return c
```

To be fair, pyright and pyrefly both choke on this, so I probably screwed something up, but I'm pretty sure these should be narrower than they are. mypy passes (possibly related due to fixing python/mypy#11685).

### Related

* facebook/pyrefly#3106
* microsoft/pyright#11377

### Version

ty 0.0.29

**Comment by carljm** [comment]
Adding a `reveal_type(source)` right after `if isinstance(source, Mapping)` clarifies what is going on here. The short version is that we are highlighting a real bug in your code.

The revealed type of `source` there is not simply `Mapping[T, SupportsInt]` (which I would guess is your expectation). The revealed type is `Mapping[T, SupportsInt] | (Iterable[tuple[T, SupportsInt]] & Top[Mapping[Unknown, object]])`. What's the second part of that union doing there? It is reflecting the possibility that an `Iterable[tuple[T, SupportsInt]]` could also be a `Mapping`! (We currently don't do a great job of simplifying that intersection type; ideally it would be simplified to `Mapping[tuple[T, SupportsInt], object]`, which would be clearer, but the result would be the same.)

So the type for `source` is the union of the two possibilities (either we have a `Mapping[T, SupportsInt]`, or we have a `Mapping[tuple[T, SupportsInt], object]`), which means the type for `outcome` is the union `T | tuple[T, SupportsInt]`, and the type for `count` is the union `SupportsInt | object`, which simplifies to `object`.

You can easily see this bug happen at runtime. Pass a dictionary with keys that are `tuple[T, SupportsInt]` and values that are _not_ `SupportsInt` to this function. That [is valid and will pass type-checking](https://play.ty.dev/b07132d0-6382-4148-a564-ccc7fabd74fa), because that dictionary is an `Iterable[tuple[T, SupportsInt]]` (dictionaries, and other mappings, are iterables of their key type.) But that dictionary will also pass `if isinstance(source, Mapping):`. And that same code will blow up at runtime, in exactly the way ty predicted:

```
➜ python3 issue3249.py
Traceback (most recent call last):
  File "/Users/carlmeyer/projects/ruff/issue3249.py", line 32, in <module>
    count_the_things(d)
    ~~~~~~~~~~~~~~~~^^^
  File "/Users/carlmeyer/projects/ruff/issue3249.py", line 20, in count_the_things
    c[outcome] += int(count)
                  ~~~^^^^^^^
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'DoesNotSupportInt'
```

Contents of `issue3249.py` are the same as the playground link above:

```py
from collections import Counter
from collections.abc import ItemsView, Iterable, Mapping
from typing import SupportsInt, TypeVar, cast, reveal_type

T = TypeVar("T")

def count_the_things(
    *sources: Mapping[T, SupportsInt] | Iterable[tuple[T, SupportsInt]],
    use_keys_instead_of_items_for_maps: bool = False,
) -> Counter[T]:
    c: Counter[T] = Counter()
    for source in sources:
        if isinstance(source, Mapping):
            if use_keys_instead_of_items_for_maps:
                for outcome in source:
                    count = source[outcome]
                    c[outcome] += int(count)
            else:
                for outcome, count in source.items():
                    c[outcome] += int(count)
        else:
            reveal_type(source)
            for outcome, count in source:
                reveal_type(outcome)
                reveal_type(count)
                c[outcome] += int(count)
    return c

class DoesNotSupportInt: ...

d = {("foo", 1): DoesNotSupportInt()}
count_the_things(d)
```

Closing as duplicate of #1578, since it's effectively the same issue, although in this case there's no multiple inheritance needed, since `Mapping` is already a subtype of `Iterable`.

**Comment by posita** [comment]
Boy do I feel like a dummy. I should have spotted that. Thanks @carljm! Apologies for taking up your time and attention with this.

**Comment by carljm** [comment]
No worries!
